csrc/transform_replay.cpp

```diff
@@ -331,40 +335,81 @@ void TransformReplay::selfReplay(
 
     std::vector<IterDomain*> new_allocation;
     std::vector<std::optional<bool>> new_contiguities;
-    new_allocation.reserve(allocation.size());
-    new_contiguities.reserve(contiguities.size());
-
-    // Push back the reduction IDs that are not mapped
-    if (ignore_reductions) {
-      for (auto* id : new_self->logical()) {
-        if (id->isReduction()) {
-          new_allocation.push_back(id);
-          new_contiguities.push_back(std::nullopt);
-        }
+
+    for (auto* new_id : new_self->logical()) {
+      if (mapped_new_ids.count(new_id) == 0) {
+        NVF_ERROR(
+            new_id->isReduction(),
+            new_id->toString(),
+            " should be a reduction.");
+        new_allocation.push_back(new_id);
+        new_contiguities.push_back(std::nullopt);
       }
     }
 
     // Pushing the mapped IDs and corresponding contiguity flags
-    for (auto&& [alloc_id, contiguity] : zip(allocation, contiguities)) {
-      if (ignore_reductions && alloc_id->isReduction()) {
+    for (const auto& [alloc_id, contiguity] : zip(allocation, contiguities)) {
+      IterDomain* new_alloc_id = getOrDefault(replay, alloc_id);
+      if (new_alloc_id == nullptr) {
         continue;
       }
-      auto it = replay.getReplay().find(alloc_id);
-      NVF_ERROR(
-          it != replay.getReplay().end(),
-          "failed to replay IterDomain: ",
-          alloc_id);
-      NVF_ERROR_EQ(
-          (it->second->isBroadcast() || it->second->isReduction()),
-          !contiguity.has_value(),
-          "Contiguity should be nullopt iff broadcast or reduction, true/false "
-          "otherwise.");
-      new_contiguities.push_back(contiguity);
-      it->second->parallelize(alloc_id->getParallelType());
-      new_allocation.push_back(it->second);
+      new_alloc_id->parallelize(alloc_id->getParallelType());
+      new_allocation.push_back(new_alloc_id);
+
+      std::optional<bool> new_contiguity = contiguity;
+      if (new_alloc_id->isBroadcast() || new_alloc_id->isReduction()) {
+        new_contiguity = std::nullopt;
+      }
+      new_contiguities.push_back(new_contiguity);
     }
 
     new_self->setAllocationDomain(new_allocation, new_contiguities);
+  } else {
```

Comment:
This feels a little unexpected to me because, even though there's nothing to replay for the allocation, the contiguity could be modified.